torch_xla/csrc/tensor_util.cpp

```diff
@@ -685,6 +708,11 @@ torch::lazy::BackendDataPtr TensorToXlaData(
     // called.
     populate_mwait->Wait();
     return async->async_datas.front();
+  } else if (ShardingUtil::UseVirtualDevice()) {
+    std::vector<std::string> local_devices =
+        xla::ComputationClient::Get()->GetLocalDevices();
+    return WrapXlaData(TransferToServerReplicated(tensor, device.toString(),
```

Comment:
It looks like we unconditionally replicate all the data? I thought tiled data shouldn't be replicated?

Reply:
In this codepath, the tensor is unsharded, and data is being immediately created on a physical device (i.e. bypassing virtual device). This data should be replicated to all devices. This is the path taken by `IrValueFromScalar`

Reply:
FYI I just moved this condition above the `transfer_async && tensor.dim() == 0 && tensor.numel() == 1` condition - PjRt always transfers async, and we should give the replicated transfer precedence

Reply:
It looks like there are a lot of places calling `TensorToXlaData`. You mentioned you patched this code path because of a case where the tensor created bypasses the virtual device and then you check UseVirtualDevice before executing the code?

Reply:
Right, this case handles the scenario where `TensorToXlaData` is called for a tensor whose device isn't `DeviceType::SPMD`. Tensors created through [UpdateFromTensor](https://github.com/pytorch/xla/blob/72d2874d83188661325886b32977782478abb47d/torch_xla/csrc/tensor.cpp#L499) will be placed on the virtual device and their data not moved off of the CPU until the `CreateTensorsData` call, and `TensorToXlaData` just returns a placeholder for them here. (cc @steventk-g)

If the virtual device is active and the tensor isn't following the `UpdateFromTensor` path, we want to implicitly replicate it here. In GPT-2 I saw this for scalars in `IrValueFromScalar`, but there may be other scenarios.

Let me know if this explanation makes sense or not, I'd be happy to adjust the approach.

Reply:
Got it. Commented on the new code.